### PR TITLE
feat: add support for custom unary, binary and ternary operators

### PR DESCRIPTION
**Pull request by hstanford**
Adds custom methods to support operators that currently have no explicit method in node-sql-ts.

The top-level `sql` instance provides methods allowing the user to create operator functions.
```
sql.prefixUnaryOperator(operator: string) => (arg: any) => Node
sql.postfixUnaryOperator(operator: string) => (arg: any) => Node
sql.binaryOperator(operator: string) => (left: any, right: any) => Node
sql.ternaryOperator(operator, separator) => (left: any, middle: any, right: any) => Node
```

Example Usage:

```
const tsMatches = sql.binaryOperator('<->');
users.select().where(tsMatches(users.biography, 'design'));
```

Alternatively, these operators are available on mixed in nodes such as columns:

```
node.prefixUnaryOperator(operator: string) => Node
node.postfixUnaryOperator(operator: string) => Node
node.binaryOperator(operator: string, right: any) => Node
node.prefixUnaryOperator(operator: string, middle: any, separator: string, right: any) => Node
```

E.g.
```
users.select().where(users.biography.binaryOperator('<->', 'design'))
```

### CODE REVIEW

**Comment by hstanford** [comment]
Thanks for converting this repo to typescript, massive amount of really valuable work! This is a rewrite of a PR I submitted to the base repo back in 2017, never got pulled as it went unmaintained.

I wanted to check about linting and prettier - I didn't end up running prettier on valueExpression.ts as it changed lines I hadn't touched. What would you like me to do here?

I also had trouble getting typescript to respect the `this` binding on the mixin, so I opted for explicit type guards instead. Hope that's ok!

**Comment by charsleysa** [comment]
Hi @hstanford 

Thanks for the contribution and sorry for not getting to this sooner!

I'd just like to clarify the use case for having a generic function like this vs adding more named operator functions.

Also, I'm not a fan of the `op` function returning another function that you need to call, I think it would be better if it could all be done as a single call, maybe using arrays like the `case` function.

Please let me know your thoughts.

**Comment by hstanford** [comment]
No worries, thanks for the response!

My background is postgres, which has a large number of operators which aren't shared with other dialects. There is also support for creating your own operators, so we will never be able to cover all of them as named operator functions in this library. 

I think eventually it would be nice for consumers of this library to have the ability to extend the functions and operators it supports when it's first loaded in, and then not declare these at "runtime". `sql.addFunction`/`sql.addOperator` or similar. Unfortunately, the way the operators are mixed in makes this very difficult (even more so with typescript!), and I thought the design in this PR might bridge the gap.

My thinking when making the `op` function return another function was that I'd want `col.equals` to behave exactly the same as `col.op('=')`. It would line up with the behaviour of `sql.function`, which I think's appropriate considering that operators are just syntactic sugar for functions. It might also allow a user to mix in their own stuff - e.g. something along the lines of `Column.prototype.tsMatches = function () {return this.op('<->')}` in the pre-typescript world.

I'm ultimately happy to defer to your preference - the `case`-like syntax would still provide the functionality - but I think the design here makes it easier and more natural to read and write expressions.

**Comment by charsleysa** [comment]
Thanks for the info!

I understand about the use case of user defined operators in sql. I actually didn't know you could do that so I learned something new today!

In regards to wanting to do something like `sql.function`, I would want to expose it in a similar way instead of adding it to the value expression mixin.

E.g. `const myOp = sql.operator('binary', '||+'); myOp(column, 1);`

Then you could also do `Column.prototype.myOp = function (right) { return myOp(this, right); }`

If you're still wanting to add it to the value expression mixin, to avoid returning another function that needs to be called, how about having an op function for each type?

E.g. `prefixUnaryOp(op), postfixUnaryOp(op), binaryOp(op, right), ternaryOp(op, middle, right)`

This would allow custom operators while only using a single function call, with the added bonus of being a bit more explicit in what kind of operator it is.

**Comment by hstanford** [comment]
I really like those ideas - I'll reimplement like `const myOp = sql.operator('binary', '||+'); myOp(column, 1);`, and add the explicit operators to the mixin

**Comment by hstanford** [comment]
The new implementation is pushed and I've updated the descriptive comment above :)

File: lib/node/valueExpression.ts [hunk 1]
```diff
@@ -71,6 +82,30 @@ const ternaryMethod = (operator: string, separator: string) => {
     };
 };
 
+const prefixUnaryMethod = (operator: string) => {
+    return function(this: INodeable) {
+        return prefixUnaryOperator(operator)(this)
```
Comment (charsleysa) [inline lib/node/valueExpression.ts]: This creates a new function every time a prefix unary operator is called which would impact performance if the operator is called in a loop

File: lib/node/valueExpression.ts [hunk 2]
```diff
@@ -71,6 +82,30 @@ const ternaryMethod = (operator: string, separator: string) => {
...
+    };
+}
+
+const postfixUnaryMethod = (operator: string) => {
+    return function(this: INodeable) {
+        return postfixUnaryOperator(operator)(this)
```
Comment (charsleysa) [inline lib/node/valueExpression.ts]: This creates a new function every time a postfix unary operator is called which would impact performance if the operator is called in a loop

File: lib/node/valueExpression.ts [hunk 3]
```diff
@@ -71,6 +82,30 @@ const ternaryMethod = (operator: string, separator: string) => {
...
+    };
+}
+
+const binaryMethod = (operator: string) => {
+    return function(this: INodeable, right: any) {
+        return binaryOperator(operator)(this, right);
```
Comment (charsleysa) [inline lib/node/valueExpression.ts]: This creates a new function every time a binary operator is called which would impact performance if the operator is called in a loop

File: lib/node/valueExpression.ts [hunk 4]
```diff
@@ -71,6 +82,30 @@ const ternaryMethod = (operator: string, separator: string) => {
...
+    };
+}
+
+const ternaryMethod = (operator: string, separator: string) => {
+    return function (this: INodeable, middle: any, right: any) {
+        return ternaryOperator(operator, separator)(this, middle, right);
```
Comment (charsleysa) [inline lib/node/valueExpression.ts]: This creates a new function every time a ternary operator is called which would impact performance if the operator is called in a loop

**Review by charsleysa — CHANGES_REQUESTED** [review]
As noted in the comments, value expression operators are negatively impacted.

My recommendation would be to separate the implementations, leave the `xxMethod` functions as they were and implement the `xxOperator` value expressions similar to the `cast` and `slice` implementations.

I would even go as a far as having a separate implementation for the `sql.xxOperator` functions so that there isn't a direct dependency on value expressions.

**Comment by hstanford** [comment]
Nice catches and recommendations - I've separated the implementations so the global ones live in their own file and the mixins are implemented like cast/slice.

**Review by charsleysa — APPROVED** [review]
LGTM

**Comment by charsleysa** [comment]
Thanks for the contribution!

**Comment by charsleysa** [comment]
@hstanford this has been included in v6.2.0

**Pull request merged**